Compiler

smwhr opened the pull request:
## Checklist

- [x] The new code additions passed the tests (`npm test`).
- [x] The linter ran and found no issues (`npm run-script lint`).

## Description

Two pictures are worth a thousand lines of code : 

<img width="182" alt="image" src="https://user-images.githubusercontent.com/1090485/153631194-b441222c-bd9b-451e-b0aa-323f4fcd0f05.png">

<img width="580" alt="image" src="https://user-images.githubusercontent.com/1090485/153631079-a48c663d-37d3-431e-85aa-0d9eef0cbfbb.png">

## Thanks

To @furkle for having done the bulk of the work 2 years ago in furkleindustries/ts-ink  
To @ephread for having written the `asOrNull` function

ephread commented:
🤯  635 files, that's going to be a tough review, not gonna lie. This is incredibly exciting though, outstanding work! Everyone involved in porting the compiler, y'all rock! 👏  🎉

@smwhr is it matching Ink 1.0.0?

The tests might need to be streamlined if we're going to move to ink-proof because a large part of the tests actually came from the extraction work done for _inkgd_, which made its way into _inkjs_, then… into ink-proof. They're probably duplicated now. (I haven't checked out the PR, I'm saying this just by looking at the diff, I might be wrong). It's a larger discussion though.

We'll likely need a buffer branch.

@y-lohse I'll have a first look this weekend. (Hopefully!)

ephread commented:
I had a quick look, I'm delighted to see that you added a version of inklecate, how exciting! I tried to compile the fixtures we currently have in the project with the new JavaScript compiler. 90% of it went well, so that's awesome, but there were a few hiccups.

1. I fixed a small typo in `inklecate.ts` that prevented json files from being written to the disk (patch below - incidentally, the patch also contains changes to `compile.js` so that `inklecate.js` is used instead of `inklecate`)
2. Files containing `EXTERNAL` declarations break the compiler (listed below).
3. The JavaScript compiler doesn't add the BOM at the beginning of the files, I don't think that's an issue, but that's a slight difference in behaviour.
4. BOM aside, most files 1:1 match their counterparts compiled with the original inklecate (yay \o/) except for about 13 of them (listed below).
5. Even though the structure is different, only three of them break the tests (marked with ❌ below), so that's good news!

------

I tried to be exhaustive but I may have missed some of the offending files.

**Won't compile due to `EXTERNAL` functions**

* `bindings/external_binding.ink`
* `bindings/game_ink_back_and_forth.ink`
* `bindings/lookup_safe_or_not_with_post_glue.ink`
* `bindings/lookup_safe_or_not.ink`

**Boolean vs Integer**

* ❌ `booleans/not_one.ink`
* ❌ `evaluation/literal_unary.ink`


**Extra whitespace in ^-prefixed strings**

* `diverts/divert_in_conditional.ink`
* `diverts/same_line_divert_is_inline.ink`
* `newlines/newline_consistency_1.ink`
* `sequences/blanks_in_inline_sequences.ink`


**Slightly different structure**

* `lists/list_range.ink`
* `lists/more_list_operations.ink`
* `newlines/newlines_trimming_with_func_external_fallback.ink`
* `weaves/conditional_choice_in_weave.ink`
* `weaves/conditional_choice_in_weave_2.ink`
* `weaves/weave_within_sequence.ink`
* ❌ `inkjs/test.ink`

------

**Patch:** [0001-Use-JavaScript-inklecate-in-compile-script.patch.zip](https://github.com/y-lohse/inkjs/files/8052002/0001-Use-JavaScript-inklecate-in-compile-script.patch.zip)

------

I hope it's helpful! I'll continue reviewing a bit later, but other things are likely to be very minor.

smwhr commented:
This should match 1.0.0.  

The bulk of the code is from the first commit where I imported all of furkle's code from ts-ink. They were all literal translation from the C# code. 

First step was to make it typecheck and compile. I had to extract the function ClosestFlowBase to its own file because it caused irresolvable circular dependancies. 

Next, I made a diff between the last commit before furkle's translation (mid 2020) and master, and the sole major difference was the "Identifier" refactoring that, if I'm not mistaken, you did.

Once done, basic "Hello World !" parsed and compiled.  
From there it's just been busy work finding errors in the original translation and using `asOrNull` instead of silent casts, something the original C# code relies a lot on (I guess you already know that from the engine port).

Changes you may be interested in are in the engine part (actually I was nicely surprised that you already ported the JsonWriter).  
One thing you may not agree with : I added a hack inside of the SimpleJson class to be able to write/read floats (you'll see that I have changed the test in that regard). The hack has only one caveat : when reading the raw json, it looks for strings of the form `,123.0` so if you have that as plain text, it'll transformed in `,"123.0f"` in the final code. I know I sometimes use text that looks like function calls left to be interpreted by the player, but I actually never used float in there.

smwhr commented:
> I hope it's helpful! I'll continue reviewing a bit later, but other things are likely to be very minor.
 Very !

[x] Won't compile due to EXTERNAL functions : [16ae252](https://github.com/y-lohse/inkjs/pull/931/commits/16ae252d1d9a422c331461bae3b74504725f47c4)
[x] Boolean vs Integer : [9852ab2](https://github.com/y-lohse/inkjs/pull/931/commits/9852ab2e68cc38cbf80ed8e86f871a4fb499108d)
[x] Extra whitespace in ^-prefixed strings : [6d33b11](https://github.com/y-lohse/inkjs/pull/931/commits/6d33b11097dfbc183d51311dcf5699837b8be107)
[ ] Slightly different structure : Not sure what's wrong in them yet

ephread commented:
> Changes you may be interested in are in the engine part (actually I was nicely surprised that you already ported the JsonWriter).

Haha, we had to! Have a look at #363 and #493 if you want to read about the rational.

> One thing you may not agree with : I added a hack inside of the SimpleJson class to be able to write/read floats (you'll see that I have changed the test in that regard).

It took me a while to wrap my head around it, but I think it's a fantastic idea! @y-lohse will call the shots on what can be added or not, but here are my thoughts on the topic.

1. I feel we should leverage the compiler to detect floats and mark them as such in the bytecode. That means the bytecode would become incompatible with the C#-runtime, but we could easily solve that problem by having a feature flag. The compiler would then output either an inkjs-compatible bytecode or a C#-compatible bytecode.
    * I feel it's okay to diverge slightly from upstream, as long as we can have an end-to-end JavaScript experience. I feel that most web users will eventually start compiling their stories with the JavaScript compiler when they are exported for the web (it's easy to include those changes in Inky).
2. On the runtime side, same story, the float conversion should be hidden away behind a feature flag, regardless of what "bytcode" is provided. That way if it messes up somebody's story, they can disable it. The runtime could also have an "inkjs" mode, where the conversions have been performed by the compiler and the saves would be incompatible with the C# runtime.

I'll open a separate issue to discuss the topic of a specific bytecode and the feature flagging. It's out of the scope of this PR. I'm going to retarget the PR to another branch name `compiler`, that way we'll be able to iterate through different PRs and keep things neat and tidy!

There's also a larger discussion to have about `inklecate.js`, for which I'll also create a issue. I'd be inclined to name it differently (maybe "nodlecate" 😛), because I don't think we should try to replicate inklecate, rather, we should improve upon it! I have a bunch a pain points it could solve.

Then, I'll create a third issue about ink-proof and how we want to move forward with it.

I'm only going to review the runtime stuff for now, because I'm not sure it would make sense to review the compiler while bugs are still being ironed out.

smwhr, I'll be the patronizing parent and urge you to rest now, we all agree that you earned a well-deserved rest this weekend! 🙃

ephread reviewed (commented):
First review of the runtime changes!

smwhr commented on src/engine/Container.ts:
```diff
@@ -180,18 +180,18 @@ export class Container extends InkObject implements INamedContent {
   }
   public AddContentsOfContainer(otherContainer: Container) {
-    this.content = this.content.concat(otherContainer.content);
+    this.content.push(...otherContainer.content);
```
Learnt in the process that concat loses references.

smwhr commented on src/engine/INamedContent.ts:
```diff
@@ -1,4 +1,4 @@
 export interface INamedContent {
-  name: string;
-  hasValidName: boolean;
+  name: string | null;
+  hasValidName?: boolean;
```
The ParsedHierarchy class that implements INamedContent never redeclare `hasValidName` it (or even use it for that matter) so this was the easy way. I'm not familiar enough with C# to know why it allows the interface to be only partially implemented 🤡

ephread replied:
Alright, I looked at the code in upstream; the two types are different but since we don't currently have namespaces in the project, I guess that works for now. Maybe we should consider putting the runtime and the compiler into different namespaces in the future!

smwhr commented:
I'm going to listen to you and enjoy the rest week-end as it is sunny here in Normandy.   
I'll be back around here in a few weeks as I may have _actual work to do_ in the mean time :)

FTR : 3 remaining ink.json are different

- `weaves/conditional_choice_in_weave.ink.json` :
```diff
- {"->":"a_stitch"},"\n",{"#f":5}]}]}],"nop","\n"
+ {"->":"a_stitch"},"\n",{"->":"0.g-0.g-1"},{"#f":5}]}]}],"nop","\n"
```

- `weaves/conditional_choice_in_weave_2.ink.json`
```diff
-[{"#n":"$r2"}],"end","\n",{"#f":5}]}]}],"nop","\n",["^bottom gather"
+[{"#n":"$r2"}],"end","\n",{"->":"0.g-1.g-2"},{"#f":5}]}]}],"nop","\n",["^bottom gather"
```

- `weaves/conditional_choice_in_weave_2.ink.json`
```diff
- "^nextline","\n","end",{"#f":5}]}],"#f":5}],"\n"
+"^nextline","\n","end",{"->":"0.g-0"},{"#f":5}]}],"#f":5}],"\n"
```

but this is the same problem in all of them as you can see.

I'll follow the discussion about the tech directions to give to this in the meantime !

ephread commented:
I had originally planned to open a specific issue about ink-proof, but the more I think about it, the more I believe that it isn't the right tool for _inkjs_. I think it complicates the infrastructure.

1. Adding ink-proof's test suite means that, whenever a new version of *ink* is released, we need to wait for ink-proof to implement or update the relevant test cases; it's not ideal.
2. Since ink-proof is purely a conformance testing tool that acts in a black box fashion, it doesn't cover everything; in particular, _inkjs_ has tests regarding function bindings and variable manipulation.
3. While we could use ink-proof for conformance tests and keep our current infrastructure for other tests, I feel it would be confusing to mix them together.

Instead of using ink-proof, I'd recommend porting the missing (compiler-only) tests from [Tests.cs](https://github.com/inkle/ink/blob/master/tests/Tests.cs). I'm thinking about compiling the ink fixtures on the fly rather than relying on the individual pre-compiled files, much like what is done in upstream. That said, I wouldn't get rid of the individual files, even if the I/O makes the tests slower. I believe we should test that the generated bytecode is the same between inklecate and the JavaScript compiler (BOM aside), hence we could keep the pre-compiled files in the repo and diff them with the result of the JavaScript compiler. Or maybe we could run ink-proof in the CI to that purpose, dunno.

Happy to discuss this further. Let me all know what you think.

I'll wait for feedback, but I'm ready to create a PR in @smwhr's fork with the tests ported from upstream. (Alternatively, we could merge the PR into `compiler` and work from there.)

------

**Note 1:** I wouldn't want anyone to think that I'm dunking on ink-proof here, 😅  it's a fantastic tool that gives a great idea of what's compatible with what.
**Note 2:** I know @smwhr must have spent a lot of time writing `proof.ts`, so whether or not we move forward with ink-proof, I want to underline once more that all the work poured in _inkjs_ is very much appreciated.

smwhr commented:
I have absolutely no sentimental value regarding `proof.ts` :) . This is 80% copy-paste work from the [original file](https://github.com/chromy/ink-proof/blob/master/drivers/inkjs_v1.10.5.js) and just a nice script where I could put `debugger;` statement and generate a pretty and satisfying report to post on the Discord.

Inkproof test cases are generally simpler that the ones in the inkjs repo, this is why I used them. They generate JSONs than are mostly under 1000 characters which make them easily readable, comparable and thus allow easier debug of the runtime generation code. But now that the bulk of the work is done, I guess we can get rid of them.  
Testing for 1:1 equality of the generated bytecode suits me very much.   
`compile.js` can have a parameter that make it compile both (inkjs and official inklecate) for last-minute comparison.

I wrote some tests for the parsing engine in [specs/parser/Core.spec.ts](https://github.com/smwhr/inkjs/blob/compiler/src/tests/specs/parser/Core.spec.ts), some of them coming straight from `Tests.cs` (I had trouble understanding exactly how ParseUntil and Interleave behaved at first).  
And TBH I had _no faith_ porting 4.2k loc of tests : 😅  But you are _most welcome_ if you feel like you have time for it (_I_'ll be the patronizing parent afterwards).  
And I'm _pretty sure_ this is how you'll catch the last remaining bug !

Regarding the BOM : it's a side effect of the C# standard library and I don't know where I stand replicating it. An "in-between" would be to have `inklenode` (another suggestion for the name, I think we should keep `ink` in it. Other suggestion include `inklets`, `inklescript`... we should vote) generate it when you write an ink.json file to disk, but not as a direct output of `Compiler.Compile` (simplifying js-js pipeline, but allowing compat if a team has js tooling for authors but subsequent pipelining the json file to unity etc.)

ephread commented:
> This is 80% copy-paste work from the [original file](https://github.com/chromy/ink-proof/blob/master/drivers/inkjs_v1.10.5.js)

Pheeew! Glad to hear you didn't spend too much time on it.

> And TBH I had no faith porting 4.2k loc of tests : 😅 But you are most welcome if you feel like you have time for it (I'll be the patronizing parent afterwards).

Haha, for sure!

We ought to do a better job at explaining the test infrastructure to contributors. I just checked CONTRIBUTING.md and _Porting tests from the reference implementation_ is… empty.

About 90% of inkjs' tests come from [Tests.cs](https://github.com/inkle/ink/blob/master/tests/Tests.cs). All of the original runtime tests were ported a few years ago, while about 20 or so compiler-only tests were tossed aside. As a result, porting the remaining tests is not as tedious as it sounds (which is why I'm more than happy to it 😎).

A substantial part of ink-proof comes from Test.cs as well, so ink-proof and inkjs share a lot in common. For instance `I104` is the same test as `multi_thread` in inkjs. In the end we wouldn't get rid of ink-proof, we'd merely have the same tests written in a slightly different fashion (except for a couple of custom-made tests).

That said, I agree, ink-proof's approach is definitely simpler, because it doesn't require porting the tests themselves. Given the mountain of work that porting the compiler involved, you made the right call.

> Testing for 1:1 equality of the generated bytecode suits me very much.
compile.js can have a parameter that make it compile both (inkjs and official inklecate) for last-minute comparison.

Awesome, we'll make that happen!

> I wrote some tests for the parsing engine in [specs/parser/Core.spec.ts](https://github.com/smwhr/inkjs/blob/compiler/src/tests/specs/parser/Core.spec.ts), some of them coming straight from Tests.cs (I had trouble understanding exactly how ParseUntil and Interleave behaved at first).

👍 I'll probably refactor them when porting over the original tests then, and we'll keep the custom ones!

> And I'm pretty sure this is how you'll catch the last remaining bug !

Let's hope so!

> […] generate it when you write an ink.json file to disk, but not as a direct output of Compiler.Compile (simplifying js-js pipeline, but allowing compat if a team has js tooling for authors but subsequent pipelining the json file to unity etc.)

I think it's a great idea, that way we have the best of both worlds. I personally think the BOM is unnecessary since everything is UTF-8 anyway. We can inject it when diffing between `inklecate`'s output and `inklenode`'s output if necessary.

> inklenode (another suggestion for the name, I think we should keep ink in it. Other suggestion include inklets, inklescript... we should vote)

Absolutely, I have a yet to create the issue, hopeful I'll get around it tomorrow and then we can start polling. I really like `inklenode` by the way!

------

I think we can pre-merge everything into `compiler` and work from there. I'll create a new PR to track the changes to the tests and then we can keep reviewing. @y-lohse I sumon you, whaddaya think?

smwhr commented:
> I'll be back around here in a few weeks as I may have actual work to do in the mean time :)

I lied.  
I found the last remaining bug. ([4d1e3c8](https://github.com/y-lohse/inkjs/pull/931/commits/4d1e3c89ca43d4bd9f89d4e2eb8097ad1b4edfaf))

I also added the BOM to inklecate.
Now if you `node compile.js`, the result give no (0, zero, nada) change 🎉

_Now_, I can go think of something else !

y-lohse commented:
Quick extra points (the answer is probably in the diff too, sorry 🙈)

- How is the porting done in the end? Is it C# translation style, or a more "liberal interpretation"? 😁 
- Do you have any thoughts about how we would ship this? I think shipping everything in one package is probably the easiest and has no big downsides, but i'm curious to see if the impact on bundle size is significant or not. I also have no idea if inkjs is tree-shakable etc.

ephread commented:
Quick note about the license, I don't think we should remove you @y-lohse, rather, we should add inkle back on top of the file. You own the original port, inkle owns the original code.

Something like:

```
Copyright (c) 2017 inkle Ltd.
Copyright (c) 2017 Yannick Lohse for the inkjs Player
Copyright (c) 2022 Julien Zamor + Furkle for the inkjs Compiler
```

We should (probably) also acknowledge other minor contributors one way or another, but I'm not an expert on copyright laws and licenses.

y-lohse commented:
That works for me too, but same, not a copyright expert at all. If we keep a copyright notice I think you should be included in there too? But ok we can discuss this separately, it's not really related to the compiler :D

smwhr commented:
> - How is the porting done in the end? Is it C# translation style, or a more "liberal interpretation"? 😁 

C# translation. I initially made good progress porting the parser to "new style" but subsequently porting the runtime generator was too much. Then I discovered furkle's code who already had made the bulk of the litteral translation.


> - Do you have any thoughts about how we would ship this? I think shipping everything in one package is probably the easiest and has no big downsides, but i'm curious to see if the impact on bundle size is significant or not. I also have no idea if inkjs is tree-shakable etc.

I modified the rollup bundler and added
- inkjs-full.js (about twice the size of inkjs.js) compiled for web that exposes the Compiler and the regular Story
- inklecate.js (soon to be renamed inklenode or something else) which is compiled targeting cjs for node execution (size not really an issue for this one, but it is around the same size as inkjs-full)
- inkjs, inkjs-e2015 and inkjs-es6 are unchanged (the Compiler never being referenced in the runtime, it's not included in those bundles)

ephread commented:
> But regardless — IMO we can merge this PR as is and work on that incrementally.

Merging then! I won't squash for now, as the history might come handy during the refinement phase. I'll create a new PR to track the remaining tasks.
